File: go/internal/profanity/profanity.go
// Package profanity calls the mock profanity-check stub with a timeout and
// fails open: any error or timeout is treated as "clean" and logged, so a
// down dependency never blocks a user's write (plan.md criterion #6).
package profanity

import (
	"bytes"
	"context"
	"encoding/json"
	"log/slog"
	"net/http"
	"time"
)

type Checker struct {
	URL     string
	Timeout time.Duration
	Client  *http.Client
}

func New(url string, timeout time.Duration) *Checker {
	return &Checker{URL: url, Timeout: timeout, Client: &http.Client{}}
}

func (c *Checker) IsClean(ctx context.Context, text string) bool {
	ctx, cancel := context.WithTimeout(ctx, c.Timeout)
	defer cancel()

	body, _ := json.Marshal(map[string]string{"text": text})
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.URL+"/check", bytes.NewReader(body))
	if err != nil {
		slog.Warn("profanity check request build failed, allowing content", "err", err)
		return true
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.Client.Do(req)
	if err != nil {
		slog.Warn("profanity check unreachable, allowing content", "err", err)
		return true
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		slog.Warn("profanity check returned non-200, allowing content", "status", resp.StatusCode)
		return true
	}

	var result struct {
		Clean bool `json:"clean"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		slog.Warn("profanity check response unreadable, allowing content", "err", err)
		return true
	}
	return result.Clean
}

File: rust/src/profanity.rs
// Calls the mock profanity-check stub with a timeout and fails open: any
// error or timeout is treated as "clean" and logged, so a down dependency
// never blocks a user's write (plan.md criterion #6).
use serde::{Deserialize, Serialize};
use std::time::Duration;

#[derive(Clone)]
pub struct Checker {
    url: String,
    timeout: Duration,
    client: reqwest::Client,
}

#[derive(Serialize)]
struct CheckRequest<'a> {
    text: &'a str,
}

#[derive(Deserialize)]
struct CheckResponse {
    clean: bool,
}

impl Checker {
    pub fn new(url: String, timeout: Duration) -> Self {
        Checker {
            url,
            timeout,
            client: reqwest::Client::new(),
        }
    }

    pub async fn is_clean(&self, text: &str) -> bool {
        let resp = self
            .client
            .post(format!("{}/check", self.url))
            .timeout(self.timeout)
            .json(&CheckRequest { text })
            .send()
            .await;

        let resp = match resp {
            Ok(r) => r,
            Err(e) => {
                tracing::warn!(err = %e, "profanity check unreachable, allowing content");
                return true;
            }
        };

        if !resp.status().is_success() {
            tracing::warn!(status = %resp.status(), "profanity check returned non-200, allowing content");
            return true;
        }

        match resp.json::<CheckResponse>().await {
            Ok(body) => body.clean,
            Err(e) => {
                tracing::warn!(err = %e, "profanity check response unreadable, allowing content");
                true
            }
        }
    }
}
